### PR TITLE
Add support for Qwen3-MoE Model

### PR DESCRIPTION
**Pull request by parambole**
### TL;DR

-   **What**: This PR integrates the **Qwen3-MoE** architecture into MaxText, with a specific implementation for the **`Qwen3-235B-A22B`** model. It includes a new MoE decoder layer and a dedicated checkpoint conversion script.

-   **Why**: To enable the pre-training of the open-source Qwen3 family of Mixture-of-Experts models within the MaxText framework.

-   **How**: By introducing a `Qwen3MoeDecoderLayer`, adding a `qwen3-235b-a22b.yml` configuration file, implementing a `norm_topk_prob` flag for Qwen-specific router logic, and creating a new `convert_qwen3_moe.py` script to convert Hugging Face checkpoints.

* * * * *

#### Background and Motivation

This pull request adds support for the Qwen3-MoE family of models, as introduced in the technical report: **[Qwen3: A More Powerful and General Base Model Series](https://arxiv.org/abs/2505.09388)**. This initial integration focuses on the `Qwen3-235B-A22B` variant, a large-scale Mixture-of-Experts (MoE) model.

#### Architectural Changes

To support the Qwen3-MoE architecture, the following key changes have been implemented:

1.  **Qwen3 Decoder Layers (`MaxText/layers/qwen3.py`)**:

    -   The existing Qwen3 module has been refactored to support both dense and MoE variants.
    -   A `self_attention_with_norm` helper function was extracted to share the attention block logic between both decoder types.
    -   A new **`Qwen3MoeDecoderLayer`** has been added. It utilizes the shared self-attention block and integrates with the existing `moe.get_routed_moe` function for the expert layers.

2.  **MoE Router Normalization (`MaxText/layers/moe.py`)**:

    -   A new boolean configuration flag, **`norm_topk_prob`**, has been added.
    -   When enabled, this flag applies a Qwen-specific normalization to the router probabilities by dividing the softmax output by its sum. This is essential for matching the original model's behavior.

3.  **New Model Configuration (`MaxText/configs/models/qwen3-235b-a22b.yml`)**:

    -   A new YAML configuration file is provided to define the specific architectural parameters for the `Qwen3-235B-A22B` model, including `num_experts`, `num_experts_per_tok`, `base_moe_mlp_dim`, and `use_qk_norm`.

4.  **Checkpoint Conversion Script (`MaxText/convert_qwen3_moe.py`)**:

    -   A new, standalone conversion script has been added to convert official Hugging Face checkpoints for the `Qwen3-235B-A22B` model into the MaxText format.
    -   This script handles the mapping of weights, including the complex stacking of MoE layers, and performs the necessary transpositions to make them compatible with MaxText's scanned layer format.

5.  **Framework Integration (`MaxText/common_types.py`, `MaxText/layers/decoders.py`)**:

    -   The `DecoderBlockType` enum has been updated with `QWEN3_MOE`.
    -   The decoder factory in `decoders.py` has been updated to correctly instantiate `Qwen3MoeDecoderLayer` when `decoder_block: "qwen3_moe"` is set in the config.

* * * * *

#### 1\. Convert Hugging Face Checkpoint

First, convert the original Hugging Face checkpoint to the MaxText format using the new script.

```
python3 -m MaxText.convert_qwen3_moe\
  --base_model_path /path/to/hf/Qwen3-235B-A22B-Thinking-2507\
  --maxtext_model_path gs://your-gcs-bucket/qwen3/scanned/step_0/\
  --model_size qwen3-235b-a22b

```

#### 2\. Run Training

Once the checkpoint is converted, you can run  training workloads by specifying the model name and loading the converted weights.

Bash

```
# Example for running validation
JAX_PLATFORMS=cpu python3 -m MaxText.tests.forward_pass_logit_checker MaxText/configs/base.yml\
  tokenizer_path=assets/qwen3-tokenizer\
  load_parameters_path=gs://your-gcs-bucket/qwen3/scanned/step_0/items/\
  model_name=qwen3-235b-a22b\
  --hf_model_path=Qwen/Qwen3-235B-A22B-Thinking-2507

```

* * * * *

### Testing Strategy & Validation

The correctness of the implementation and the conversion script was validated by comparing the logits produced by the MaxText model against the original Hugging Face model (`Qwen/Qwen3-235B-A22B-Thinking-2507`).

The `forward_pass_logit_checker` test was used with several prompts.

**Key Validation Metrics:**

| Prompt | Overlap Count | Jaccard Similarity | Rank Agreement % | Avg KL Divergence |
| --- | --- | --- | --- | --- |
| "I love to" | **10/10** | **1.0** | 40.0% | 0.000254 |
| "Today is a" | **10/10** | **1.0** | 80.0% | 0.001808 |
| "What is the" | **10/10** | **1.0** | 100.0% | -0.000182 |

* * * * *

### Checklist

Before submitting this PR, please make sure (put X in square brackets):

-   [X] I have performed a self-review of my code.
-   [X] I have necessary comments in my code, particularly in hard-to-understand areas.
-   [X] I have run end-to-end tests and provided the validation results above.
-   [X] I have made or will make corresponding changes to the documentation if needed.

### CODE REVIEW

File: MaxText/layers/qwen3.py [hunk 1]
```diff
@@ -36,9 +36,90 @@
...
+      max_prefill_predict_length=cfg.max_prefill_predict_length,
+      attention_kernel=cfg.attention,
+      mesh=mesh,
+      dtype=cfg.dtype,
+      weight_dtype=cfg.weight_dtype,
+      dropout_rate=cfg.dropout_rate,
+      name="self_attention",
+      quant=quant,
+      kv_quant=quantizations.configure_kv_quant(cfg),
+      use_qk_norm=True,
```
Comment (shralex) [inline MaxText/layers/qwen3.py]: lets use cfg.use_qk_norm here, its provided by the configs

File: MaxText/convert_qwen3_moe.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
+"""
```
Comment (gagika) [inline MaxText/convert_qwen3_moe.py:1]: instead of this, could you please integrate the Qwen-MoE logic into the existing utility functions in MaxText/utils/ckpt_conversion/utils/?

The QWEN3_MAXTEXT_TO_HF_PARAM_MAPPING function already handles the dense Qwen architecture. Extending it to also cover the MoE variant would be ideal. This is a key step toward enabling direct loading (or saving) of Hugging Face checkpoints in the future.

Comment (parambole) [reply MaxText/convert_qwen3_moe.py:1]: Hey @gagika,

Originally the plan was to use the new `Util` for checkpoint conversion and was part of this PR. But I was facing issues while using the new `Util` for converting checkpoints specifically the script 
1. Crashed while converting the checkpoints to the current model size
2. The MoE specific layer support was missing ( which I later added )

Due to the above hiccups I separated it into a different PR https://github.com/AI-Hypercomputer/maxtext/pull/2149 so that I can get this PR unblocked. I will work with @hengtaoguo and @YixuanWang-99 to better understand the issues and use PR#2149 to converge. 

Thanks!

File: MaxText/configs/models/qwen3-235b-a22b.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+base_num_kv_heads: 4
+base_num_decoder_layers: 94
+head_dim: 128
+mlp_activations: ["silu", "linear"]
+vocab_size: 151936
+normalization_layer_epsilon: 1.0e-6
+use_qk_norm: True
+
+# MoE Specific Parameters
+decoder_block: "qwen3_moe"
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-235b-a22b.yml]: Nit: probably it's better to move to architecture param? decoder_block indicates the whole layer, not specifically to MoE/MLP block.

File: MaxText/convert_qwen3_moe.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+import numpy as np
+import torch
+from safetensors import safe_open
+from tqdm import tqdm
+
+from MaxText import llama_or_mistral_ckpt, max_logging
+from MaxText.inference_utils import str2bool
+
+# Static model parameters dictionary
+MODEL_PARAMS_DICT = {
```
Comment (RissyRan) [inline MaxText/convert_qwen3_moe.py:39]: Oh, yes. If we could use param mapping instead of the script. It will be great!

File: MaxText/layers/qwen3.py [hunk 5]
```diff
@@ -52,106 +136,101 @@ def __call__(
...
+    if cfg.scan_layers:
+      return layer_output, None
+    else:
+      return layer_output
+
+
+# -----------------------------------------
+# The MoE Decoder Layer for Qwen3
+# -----------------------------------------
+class Qwen3MoeDecoderLayer(nn.Module):
```
Comment (RissyRan) [inline MaxText/layers/qwen3.py:186]: Do you know if we could directly use nnx.module? otherwise, we will migrate it again. One example here: https://github.com/AI-Hypercomputer/maxtext/compare/main...gpt_layer_nnx. However, I met some memory overheads. cc @bvandermoon @cgarciae if you have some ideas.

But I am totally fine if we cannot refactor to nnx at this moment.

File: end_to_end/tpu/qwen/moe/qwen3-235b-a22b/1_test_qwen3_235b_a22b.sh [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  sparse_matmul=False \
+  load_parameters_path=${MAXTEXT_CHECKPOINT_PATH} \
+  model_name=qwen3-235b-a22b \
+  checkpoint_storage_concurrent_gb=512 \
+  skip_jax_distributed_system=True \
+  --hf_model_path=${HF_MODEL_PATH} \
+  --max_kl_div=0.015 \
+  --run_hf_model=True
+
+echo "Validation complete."
```
Comment (RissyRan) [inline end_to_end/tpu/qwen/moe/qwen3-235b-a22b/1_test_qwen3_235b_a22b.sh:55]: I am a little bit curious why higher divergence gives better rank agreement. Not a blocker to PR, just trying to understand the value for ranking.

https://screenshot.googleplex.com/54wgRZA3DV9a3ht

File: end_to_end/tpu/qwen/moe/run_qwen_moe.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+* * * * *
+
+Checkpoint Conversion
+---------------------
+
+To get started, you first need a MaxText-compatible checkpoint.
+
+1.  **Download the Model**: Download the official model from Hugging Face: [Qwen/Qwen3-235B-A22B-Thinking-2507](https://huggingface.co/Qwen/Qwen3-235B-A22B-Thinking-2507). You can use a tool like `hf_transfer` for a fast download.
+
+    Bash
```
Comment (RissyRan) [inline end_to_end/tpu/qwen/moe/run_qwen_moe.md]: Nit: Is this `Bash` expected? https://screenshot.googleplex.com/8wpmqbcZRCfWvV3

Comment (parambole) [reply end_to_end/tpu/qwen/moe/run_qwen_moe.md]: Done.

File: MaxText/configs/models/qwen3-235b-a22b.yml (hunk 3 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-235b-a22b.yml]: Makes sense. Done

File: end_to_end/tpu/qwen/moe/qwen3-235b-a22b/1_test_qwen3_235b_a22b.sh (hunk 6 above)
Comment (parambole) [reply end_to_end/tpu/qwen/moe/qwen3-235b-a22b/1_test_qwen3_235b_a22b.sh:55]: > I am a little bit curious why higher divergence gives better rank agreement.

That's a great question. My understanding is that the two metrics don't have a direct correlation. It seems the rank agreement can fluctuate, even with a near-zero KL divergence, because the top tokens have such close, competing probability scores. It's like a tiny numerical difference can flip their order, which impacts the rank agreement. But since the scores themselves are almost identical, the overall distribution that KL divergence measures still confirms a near-perfect match.

File: MaxText/layers/qwen3.py (hunk 5 above)
Comment (parambole) [reply MaxText/layers/qwen3.py:186]: > Do you know if we could directly use nnx.module? 

That is a great call-out. TBH I am not sure. @bvandermoon any suggestions ?

File: MaxText/convert_qwen3_moe.py (hunk 4 above)
Comment (parambole) [reply MaxText/convert_qwen3_moe.py:39]: Responded here https://github.com/AI-Hypercomputer/maxtext/pull/2092#discussion_r2267717951

Thanks!

**Comment by parambole** [comment]
> Quick double check, the fine-tuning end-to-end is verified right?

@RissyRan Can you please elaborate ?

File: MaxText/layers/qwen3.py (hunk 5 above)
Comment (bvandermoon) [reply MaxText/layers/qwen3.py:186]: @RissyRan where did you see memory overheads? We are currently debugging additional memory during inference

File: end_to_end/tpu/qwen/moe/run_qwen_moe.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    run_name=qwen3_finetuning\
+    per_device_batch_size=4\
+    model_name=qwen3-235b-a22b\
+    steps=500\
+    max_target_length=2048\
+    ici_fsdp_parallelism=128\
+    attention=flash\
+    tokenizer_path=assets/qwen3-tokenizer\
+    # Qwen3-specific flags
+    decoder_block="qwen3_moe"\
```
Comment (gobbleturk) [inline end_to_end/tpu/qwen/moe/run_qwen_moe.md]: The Qwen3 specific flags should be set by `model_name=qwen3-235b-a22b` You should have 

decoder_block="qwen3_moe"
use_qk_norm=True
norm_topk_prob=True

defined in MaxText/configs/models/qwen3-235b-a22b.yml. These three are core propeties of the model, e.g. they effect the correctness of the model

Comment (parambole) [reply end_to_end/tpu/qwen/moe/run_qwen_moe.md]: Sounds good. I have updated the command.

File: MaxText/layers/qwen3.py (hunk 5 above)
Comment (RissyRan) [reply MaxText/layers/qwen3.py:186]: @bvandermoon, please check this [comment](https://b.corp.google.com/issues/437369732#comment2) for details. I probably missed something there in NNX version for training.

```
# NN:
Total memory size: 57.9 GB, Output size: 30.7 GB, Temp size: 27.2 GB, Argument size: 30.7 GB, Host temp size: 0.0 GB.

# NNX:
Total memory size: 79.8 GB, Output size: 30.7 GB, Temp size: 49.1 GB, Argument size: 30.7 GB, Host temp size: 0.0 GB.
```

@parambole If we still have issues about decoding, totally fine to push nn.module for Qwen3 at this moment.

File: MaxText/pyconfig.py [hunk 9]
```diff
@@ -796,7 +797,9 @@ def set_and_validate_pipeline_config(raw_keys):
   if using_pipeline_parallelism(raw_keys):
     # For pipeline parallelism, model_fsdp_ag_once should be False, and pipeline_fsdp_ag_once is typically True.
     if raw_keys["model_fsdp_ag_once"]:
-      raise ValueError("You should only set pipeline_fsdp_once=True, leave model_fsdp_ag_once=False with pipeline parallelism.")
+      raise ValueError(
```
Comment (RissyRan) [inline MaxText/pyconfig.py:813]: Nit: could we skip those format changes if it's not an error?

File: MaxText/pyconfig.py [hunk 10]
```diff
@@ -954,22 +957,15 @@ def validate_deepseek_moe(raw_keys):
...
-      )
+  if raw_keys["sparse_matmul"] and (
+      using_fsdp_and_transpose_parallelism(raw_keys)
+      and using_expert_parallelism(raw_keys)
+      and using_tensor_parallelism(raw_keys)
   ):
-    raise ValueError(
-        "Sparse matmul doesn't support using fsdp, expert, and tensor parallelism together."
-    )
+    raise ValueError("Sparse matmul doesn't support using fsdp, expert, and tensor parallelism together.")
```
Comment (RissyRan) [inline MaxText/pyconfig.py:981]: Could we remove this assertion? 

https://github.com/AI-Hypercomputer/maxtext/blob/c5bbad832d91eb0174406e976967b9a87134b3f8/MaxText/pyconfig.py#L212

File: MaxText/configs/models/qwen3-235b-a22b.yml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+num_experts_per_tok: 8
+base_moe_mlp_dim: 1536
+load_balance_loss_weight: 0.001
+norm_topk_prob: true
+
+# RoPE Settings
+rope_max_timescale: 5000000
+
+# General Model Settings
+enable_dropout: False
```
Comment (RissyRan) [inline MaxText/configs/models/qwen3-235b-a22b.yml]: Nit: need extra line for formatting

**Comment by RissyRan** [comment]
> > Quick double check, the fine-tuning end-to-end is verified right?
> 
> @RissyRan Can you please elaborate ?

Oh, I meant to say if you have loaded the ckpt, and run SFT for some steps?

File: MaxText/configs/models/qwen3-235b-a22b.yml (hunk 11 above)
Comment (parambole) [reply MaxText/configs/models/qwen3-235b-a22b.yml]: Done.

File: MaxText/pyconfig.py (hunk 10 above)
Comment (parambole) [reply MaxText/pyconfig.py:981]: Oh I thought I removed it. Might have gotten it back in one of the back merges from main. Thanks for catching this.

File: MaxText/pyconfig.py (hunk 9 above)
Comment (parambole) [reply MaxText/pyconfig.py:813]: Hey Ran, these were applied by the code_stlye util ( I added them as they would also show up in the future ).

**Review by gagika — APPROVED** [review]

File: MaxText/configs/base.yml [hunk 12]
```diff
@@ -169,6 +169,7 @@ use_random_routing: False # whether to use random routing for debug/test purpose
 tile_batch_seq: 512
 tile_activation_dim: 1024
 tile_weight_dim: 1024
+norm_topk_prob: False # Boolean to enable the top-k probability normalization.
```
Comment (gobbleturk) [inline MaxText/configs/base.yml]: nit: include in comment this is used with qwen moe models

File: MaxText/layers/decoders.py [hunk 13]
```diff
@@ -380,9 +381,7 @@ def move_to_device(variables):
 
           def map_fn(path, value):
             max_logging.log(f"models.py: Moving parameter {path} to device")
-            return jax.device_put(
-                value, max_utils.device_space()
-            )
+            return jax.device_put(value, jax.memory.Space.Device)
```
Comment (gobbleturk) [inline MaxText/layers/decoders.py]: oh does this work? I thought this required jax 0.7.1?

Comment (gobbleturk) [reply MaxText/layers/decoders.py]: This looks unrelated to your PR, I think this should still be max_utils.device_space()?

**Review by gobbleturk — APPROVED** [review]
Thanks Param, left one small change to make

File: MaxText/layers/decoders.py (hunk 13 above)
Comment (parambole) [reply MaxText/layers/decoders.py]: That is a awesome catch. I am not sure how that creeped it. Reverted

**Review by shralex — APPROVED** [review]

File: MaxText/layers/qwen3.py (hunk 5 above)
Comment (bvandermoon) [reply MaxText/layers/qwen3.py:186]: Thanks @RissyRan. Followed up on the bug.

For your comment around implementing this in NNX - ideally we would do that. Unfortunately we are in the process of migrating the first decoder layer currently: https://github.com/AI-Hypercomputer/maxtext/pull/2178.

When do you need this by? If we can wait for that PR to merge, then NNX makes sense here and will avoid the fast-follow migration needed (it will probably need to be migrated next week)
